### PR TITLE
Add WT_PROFILE_ID to the environment of the spawned process

### LINKED ISSUE #3589: Resolve the Profile name from the WT_SESSION ID, or publish the profile GUID as WT_SESSIONGUID
# Description of the new feature/enhancement

We need a way to detect within a running terminal what profile instantiated it, so that we can make on-the-fly changes to the config (like change the background or font). The MSTerminalSetting is an example of that method.

# Proposed technical implementation details (optional)
Provide the steps/api to PInvoke to get the profile guid based on the session guid. OR expose the session's profile guid to an environment variable like WT_PROFILE

### LINKED ISSUE #4566: Consider adding a WT_PROFILE env var pointing to settings file for 1.0 release
If the settings UI is not planned until after 1.0, then I strongly recommend adding an environment variable with the path to the json settings file, so it can be edited with the editor of choice (vi, notepad, nano, emacs, whatever) from within the shell of choice without burning brain calories trying to find it buried in AppData. Also: Relying on the GUI menu is, well, unreliable, due to there being no guarantee of a valid file association for json files.

Hmm, I'll make an issue for this independently. We've already got `WT_SESSION` env var -- something like `WT_PROFILE` would be useful. 

Then from within cmd or powershell, I can easily edit settings with my preferred editor, or using `$settings = gc $env:WT_PROFILE | convertto-json` and flipping settings on the object model directly before saving it back out.

### PR DESCRIPTION
**Pull request by oising**
This commit adds a `WT_PROFILE_ID` environment variable, which contains
the guid of the active profile.

It also teaches ConptyConnection to take an environment map on creation.

We had to do a little manual jiggery with the WSLENV environment
variable as passed by the creator.

* [x] CLA signed
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [x] I've discussed this with core contributors already.

Ran terminal, validated vars and translated paths under windows and WSL. 

References #4566 (this PR originally introduced WT_SETTINGS/DEFAULTS)
Closes #3589

### CODE REVIEW

File: OpenConsole.sln [hunk 1]
```diff
@@ -1192,6 +1180,7 @@ Global
 		{376FE273-6B84-4EB5-8B30-8DE9D21B022C}.Debug|Any CPU.Build.0 = Debug|Any CPU
 		{376FE273-6B84-4EB5-8B30-8DE9D21B022C}.Debug|ARM64.ActiveCfg = Debug|Any CPU
 		{376FE273-6B84-4EB5-8B30-8DE9D21B022C}.Debug|x64.ActiveCfg = Debug|Any CPU
+		{376FE273-6B84-4EB5-8B30-8DE9D21B022C}.Debug|x64.Build.0 = Debug|Any CPU
```
Comment (zadjii-msft) [inline OpenConsole.sln]: Why are there changes to this file at all?

File: src/cascadia/TerminalConnection/ConptyConnection.cpp [hunk 2]
```diff
@@ -10,14 +10,16 @@
 #include "ConptyConnection.h"
 
 #include <windows.h>
+//#include <winrt/Windows.Foundation.Collections.h>
```
Comment (zadjii-msft) [inline src/cascadia/TerminalConnection/ConptyConnection.cpp]: Remove dead code (and below)

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 3]
```diff
@@ -8,6 +8,7 @@
 #include "../../types/inc/utils.hpp"
 
 #include <LibraryResources.h>
+//#include <winrt/Windows.Foundation.Collections.h>
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/TerminalPage.cpp]: Oh no more dead code

File: OpenConsole.sln (hunk 1 above)
Comment (oising) [reply OpenConsole.sln]: I have no idea -- visual poodio just added them for some reason best known to itself. I only noticed this now. I will filter out the changes.

File: src/cascadia/TerminalConnection/ConptyConnection.cpp (hunk 2 above)
Comment (oising) [reply src/cascadia/TerminalConnection/ConptyConnection.cpp]: Will do. I commented and uncommented this line so many times trying to kill the squiggles of death that I lost track of it.

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 3 above)
Comment (oising) [reply src/cascadia/TerminalApp/TerminalPage.cpp]: Ditto as above.

**Comment by oising** [comment]
> Honestly this looks okay to me. Plumbing the `StringMap` through makes it easier to add some of the other requests (looking at #3589) as well.

It seems trivial to roll #3589 into this PR as well -- shall I? it's another three lines of code...

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 4]
```diff
@@ -625,10 +624,14 @@ namespace winrt::TerminalApp::implementation
 
         else
         {
+            wchar_t guidWString[40] = { 0 };
+            StringFromGUID2(profileGuid, guidWString, 40);
```
Comment (zadjii-msft) [inline src/cascadia/TerminalApp/TerminalPage.cpp]: `Utils::GuidToString` might be easier here

Comment (oising) [reply src/cascadia/TerminalApp/TerminalPage.cpp]: I would change this to please thee, but for some reason now my vs 2019 (preview) won't start "the application cannot start." It literally died after my last build. Trying to figure that out is taking longer than expected.

Comment (oising) [reply src/cascadia/TerminalApp/TerminalPage.cpp]: ok done

**Comment by JustinGrote** [comment]
I noticed WT_PROFILE is in 0.10, does there need to be a code review to make sure that change doesn't conflict with what's happening here?
image

**Comment by DHowett-MSFT** [comment]
(On account of you deleted your comment, I'm guessing something else was up :wink:)

**Comment by JustinGrote** [comment]
If it was me, I realized my comment was incorrect, it was special local config I did. All is good and looking forward to seeing this in 0.11!

**Comment by oising** [comment]
> (On account of you deleted your comment, I'm guessing something else was up 😉)

Hey @DHowett-MSFT -- did I miss something? Do I need to update the PR?

**Comment by JustinGrote** [comment]
@oising this was directed at me I believe.

**Comment by miniksa** [comment]
Not marking AutoMerge to give @DHowett-MSFT the opportunity to swing back this morning.

**Comment by oising** [comment]
> (blocking only so that I can discuss with kayla how this fits into our eventual plan to rename profiles.json -- whether we want a generic env var name, and how this will impact compatibility if people come to rely on it.)

Renamed WT_PROFILES to WT_SETTINGS, as per https://github.com/microsoft/terminal/pull/5199

**Comment by jsoref** [comment]
@DHowett-MSFT: fwiw, this is the scheduler feature running. Here, the commits were made before the spell checker was merged in (and thus, the branch in https://github.com/oising/terminal/tree/feature/profile-env-vars/.github couldn't have run the spell checker because it didn't exist).

@oising: if you merge/rebase to master, the bot will have more useful feedback. Most (possibly all?) of those errors were fixed between where you branched off and where master is today.

**Comment by jsoref** [comment]
(I could try to have the bot look up the PR merge target, try a merge, and then review, that'd increase some risks/complexity -- e.g. what to do if it hits merge conflicts. -- If we run into this a lot, I'll revisit.)

**Comment by oising** [comment]
> (I could try to have the bot look up the PR merge target, try a merge, and then review, that'd increase some risks/complexity -- e.g. what to do if it hits merge conflicts. -- If we run into this a lot, I'll revisit.)

It's okay -- I was forced to merge master anyway because another PR broke my code :P I think we're good now.

**Comment by jsoref** [comment]
(The double comment is because the schedule bot can't be certain that the other repo instance is running -- as it wasn't before. I'll be able to at least collapse one of them when I write the code to collapse older comments. -- Sadly, I haven't found an easy way to *find )

`IMap` should go into a file in:
https://github.com/microsoft/terminal/tree/master/.github/actions/spell-check/dictionary

Probably a file for `windows-foundation-api.txt` or similar (@miniksa / @DHowett-MSFT can pick a name for it).

(The other jitter isn't your fault, so I'd ignore it.)

**Comment by oising** [comment]
@jsoref Is there an action point for me here?

**Comment by jsoref** [comment]
@oising: probably create a file called `windows-foundation-api.txt` in the `dictionary` directory and add a single line with `IMap`.

**Comment by miniksa** [comment]
> Probably a file for `windows-foundation-api.txt` or similar (@miniksa Michael Niksa FTE / @DHowett-MSFT Dustin Howett FTE can pick a name for it).

All the other existing I* something names of interfaces are in `whitelist.txt` right now.
I'd just name it `apis.txt` or something if it's to go into a dictionary and not make it exclusive to `winrt::Windows::Foundation` (and maybe we'll move some of the I* things from whitelist into that eventually?)

**Comment by oising** [comment]
> > Probably a file for `windows-foundation-api.txt` or similar (@miniksa Michael Niksa FTE / @DHowett-MSFT Dustin Howett FTE can pick a name for it).
> 
> All the other existing I* something names of interfaces are in `whitelist.txt` right now.
> I'd just name it `apis.txt` or something if it's to go into a dictionary and not make it exclusive to `winrt::Windows::Foundation` (and maybe we'll move some of the I* things from whitelist into that eventually?)

Added apis.txt.

**Comment by oising** [comment]
> #### New misspellings found, please review:
> * CHINESEBIG
> * IMap
> 
> To accept these changes, run the following commands
>  ✏️ Contributor please read this
> #### ⚠️ Reviewers
> At present, the action that triggered this message will not show its ❌ in this PR unless the branch is within this repository.
> Thus, you **should** make sure that this comment has been addressed before encouraging the merge bot to merge this PR.

I added "IMap" to apis.txt under dictionaries. Why is this still firing? And where did "CHINESEBIG" suddenly come from?

**Comment by jsoref** [comment]
In general, if you have the spell checker on your branch, then it should use everything from your branch. There are actually two copies of the spell checker that run, one in your branch in your repository, and one from the microsoft repository. (There may be bugs in the second run, especially, it's possible I'm doing a bad job of clearing caches between checks of multiple PRs, the code is fairly new and I'm not perfect.)

The one that commented here:
https://github.com/oising/terminal/commit/2a82bde463b80caae21f0d78430f3bdb7c8f9e09
is straightforward in that it's purely your repository/your branch's view of the world. (Essentially it's unaware of pull requests / the microsoft repository / master branch.)

I need to look, if you're feeling adventurous, the commands the action took are listed in https://github.com/oising/terminal/runs/560741925?check_suite_focus=true

Hopefully I'll be able to look tonight. If not, I might not have any time for about two weeks.

Oddly, the only sign of CHINESEBIG is in a dictionary, which ideally should be treated as a "trusted" item as opposed to something to flag...

**Comment by jsoref** [comment]
@oising: 
It looks like the problem is that you didn't have a `newline` at the end of the file.

I'll have to think about how to handle this. Probably the tool should recognize the error and complain about it directly.

GitHub won't let me make a code suggestion to fix this.

As a general rule, non-Windows tools really do not approve of files that don't have a new line at the end of the file.

Fixing that resulted in the checker being happy.

**Comment by jsoref** [comment]
I've added code to the spell checker to handle this, so the next release will be nicer about it. Thanks for helping me identify this edge case 🙂.

**Comment by oising** [comment]
@DHowett-MSFT There's a persistent change request from you lingering that I can't figure out how to resolve. Can you undo it or redo it? No matter where I click, I can't find anything to address.

**Comment by zadjii-msft** [comment]
Okay I just tried to yolo the spellcheck merge, hopefully it worked?

**Comment by oising** [comment]
> Okay I just tried to yolo the spellcheck merge, hopefully it worked?

"this branch has conflicts that must be resolved."

**Comment by oising** [comment]
> I'm switching myself to make sure this doesn't get automerged into 1.0 accidentally. I believe the consensus on the team is that this is a bit too much of a "feature" to get merged this close to 1.0 tbh. I've moved it into 1.x to have the discussion once the official 1.0 build has been cut 😄

That's disappointing. You guys are shipping 1.0 without a means to edit settings from the UI, and now making it even harder to find the files if you do need to edit them.

**Comment by zadjii-msft** [comment]
I mean, <kbd>Ctrl+,</kbd> will open the settings file by default, and `%LOCALAPPDATA%\packages\Microsoft.WindowsTerminal_8wekyb3d8bbwe\LocalState\settings.json` == `%WT_SETTINGS%` for all intents and purposes. Sure it's more verbose, but it's no less correct. I get how this is useful, but we're just in a hard place where accepting new features like this late in the cycle is a little hard to sell up the chain unfortunately 😕 

Just because it's not making the 1.0 cut doesn't mean it's never happening. It's not like we're gonna ship 1.0 then all get the 'rona and disappear. We just need to make hard cuts on what actually gets in at this point unfortunately (case in point #3789). We'll certainly come back to this ☺️

**Comment by Jaykul** [comment]
In my opinion, the thing that's really missing in 1.0 is WT_PROFILE_ID which cannot be discovered at all ... @zadjii-msft 

And since the VT color _queries_ were also never finished, there's no way to discover what crazy colors someone's using in their terminal

**Comment by oising** [comment]
As per discussion with @DHowett-MSFT at https://github.com/microsoft/terminal/issues/4566#issuecomment-613102437 - removed `WT_SETTINGS` and `WT_DEFAULTS` and left `WT_PROFILE_ID`

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 5]
```diff
@@ -583,12 +590,21 @@ namespace winrt::TerminalApp::implementation
...
-                                                                    settings.StartingDirectory(),
-                                                                    settings.StartingTitle(),
-                                                                    settings.InitialRows(),
-                                                                    settings.InitialCols(),
-                                                                    winrt::guid());
+            std::wstring guidWString = Utils::GuidToString(profileGuid);
+
+            StringMap envMap{};
+            envMap.Insert(L"WT_PROFILE_ID", guidWString);
+            envMap.Insert(L"WSLENV", L"WT_PROFILE_ID");
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalApp/TerminalPage.cpp:597]: Minor concern here: WSLENV could exist in the user's preexisting environment, and we would stomp it by doing this; is there a way for us to get/append the current value like we do in ConptyConnection @ 113?

File: src/cascadia/TerminalConnection/ConptyConnection.cpp [hunk 6]
```diff
@@ -110,7 +110,17 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
...
+            {
+                // add additional WT env vars like WT_SETTINGS, WT_DEFAULTS and WT_PROFILE_ID
+                for (auto item : _environment)
+                {
+                    environment.insert_or_assign(item.Key().c_str(), item.Value().c_str());
+                }
+            }
+
             auto wslEnv = environment[L"WSLENV"]; // We always want to load something, even if it's blank.
+            //
```
Comment (DHowett-MSFT) [inline src/cascadia/TerminalConnection/ConptyConnection.cpp:123]:
```suggestion

```

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 5 above)
Comment (oising) [reply src/cascadia/TerminalApp/TerminalPage.cpp:597]: Good catch. Fixed.

File: src/cascadia/TerminalConnection/ConptyConnection.cpp [hunk 7]
```diff
@@ -108,20 +108,31 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
             const auto guidSubStr = std::wstring_view{ wsGuid }.substr(1);
 
             // Ensure every connection has the unique identifier in the environment.
-            environment.insert_or_assign(L"WT_SESSION", guidSubStr.data());
+            environment.insert_or_assign(L"WT_SESSION", guidSubStr.data());
```
Comment (jsoref) [inline src/cascadia/TerminalConnection/ConptyConnection.cpp]: You didn't want to add this trailing whitespace
```suggestion
            environment.insert_or_assign(L"WT_SESSION", guidSubStr.data());
```

**Comment by Jaykul** [comment]
Thanks for rethinking this one, y'all.

**Pull request merged**